### ISSUE TITLE
Shadows and unknown pixel-label in inference

### ISSUE THREAD

**Issue by GabrieleGalimberti-GaleSelector** [DeepSceneSeg/EfficientPS#24]
Hi
This is an output of EfficientPS without blend of original image.

![000346](https://user-images.githubusercontent.com/49124856/120114265-aac4db00-c17e-11eb-8534-abea57df9c87.png)

Why are there shadows close to cars and other dynamic elements in the image? is it an error?

Thank you

**Comment by mohan1914** [comment]
Can you provide the original image so that I can replicate the issue?

**Comment by GabrieleGalimberti-GaleSelector** [comment]
Yes, thank you... 
these shadows and black pixel assignement compare in every image with instances of KITTI.
This is the image 000346.png of KITTI odometry dataset sequence 05 camera left 

![000346](https://user-images.githubusercontent.com/49124856/121004981-98a2f800-c78f-11eb-9dd7-871b34d79e41.png)

Thank you for your reply !
Gabriele

**Comment by GabrieleGalimberti-GaleSelector** [comment]
Furthermore I don't understand if is the black part in the image result is a postprocessing result or the mask of instance is translate by something ... I really need a help.

I paste my code:

`
import argparse
import os

import mmcv
import torch
import numpy as np
import json

from mmcv.parallel import MMDataParallel, MMDistributedDataParallel
from mmcv.runner import get_dist_info, init_dist, load_checkpoint
from tools.fuse_conv_bn import fuse_module

from mmdet.apis import multi_gpu_test, single_gpu_test
from mmdet.core import wrap_fp16_model
from mmdet.datasets import build_dataloader, build_dataset
from mmdet.datasets.cityscapes import PALETTE
from mmdet.models import build_detector
from mmdet.apis import init_detector, inference_detector, show_result
from mmdet.core import cityscapes_originalIds

from skimage.morphology import dilation
from skimage.segmentation import find_boundaries

from PIL import Image

def parse_args():
    parser = argparse.ArgumentParser(
        description='MMDet test (and eval) a model')
    parser.add_argument('config', help='test config file path')
    parser.add_argument('checkpoint', help='checkpoint file')
    parser.add_argument('input', help='input folder')
    parser.add_argument('out', help='output folder')
    parser.add_argument('--local_rank', type=int, default=0)
    args = parser.parse_args()
    if 'LOCAL_RANK' not in os.environ:
        os.environ['LOCAL_RANK'] = str(args.local_rank)
    return args


def main():
    #args = parse_args()

    #model = init_detector(args.config, args.checkpoint, device='cuda:0')
    model = init_detector("efficientPS_kitti/config/efficientPS_kitti.py", "efficientPS_kitti/model/model.pth", device='cuda:0')

    images = []
    annotations = []
    if not os.path.exists("output_kitti05_PS"):
        os.mkdir("output_kitti05_PS")
    out_base = "efficientPS_TEST"#args.out.split('/')[-1]
    out_base_folder = os.path.join("output_kitti05_PS",out_base)#os.path.join(args.out, out_base)
    out_base_json = out_base_folder + '.json'

    if not os.path.exists(out_base_folder):
        os.mkdir(out_base_folder)

    PALETTE.append([0,0,0])
    colors = np.array(PALETTE, dtype=np.uint8)

    originalIds = cityscapes_originalIds()

    kitti_sequence = "../../../../media/RAIDONE/galimberti/kitti/05"
    for channel in os.listdir(kitti_sequence):
    #for channel in os.listdir("/home/galimberti/projects/EfficientPS/prova_cityscapes"):
        #path = os.path.join("/home/galimberti/projects/EfficientPS/prova_cityscapes", channel)
        path = os.path.join(kitti_sequence, channel)
        print(path)
        prog_bar = mmcv.ProgressBar(len(os.listdir(path)))

        for imgName in os.listdir(path):
            result = inference_detector(model, os.path.join(path, imgName), eval='panoptic')
            pan_pred, cat_pred, _ = result[0]
            pan_pred, cat_pred = pan_pred.numpy(), cat_pred.numpy()
            #print(pan_pred)

            imageId = imgName.replace("_leftImg8bit.png", "")
            inputFileName = imgName
            outputFileName = imgName.replace("_leftImg8bit.png", "_panoptic.png")
            # image entry, id for image is its filename without extension
            images.append({"id": imageId,
                           "width": int(pan_pred.shape[1]),
                           "height": int(pan_pred.shape[0]),
                           "file_name": inputFileName})

            pan_format = np.zeros(
                (pan_pred.shape[0], pan_pred.shape[1], 3),
                dtype=np.uint8
            )

            panPredIds = np.unique(pan_pred)
            segmInfo = []
            for panPredId in panPredIds:
                if cat_pred[panPredId] == 255:
                    continue
                elif cat_pred[panPredId] <= 10:
                    semanticId = segmentId = originalIds[cat_pred[panPredId]]
                else:
                    semanticId = originalIds[cat_pred[panPredId]]
                    segmentId = semanticId * 1000 + panPredId

                isCrowd = 0
                categoryId = semanticId

                mask = pan_pred == panPredId
                color = [segmentId % 256, segmentId // 256, segmentId // 256 // 256]
                pan_format[mask] = color

                area = np.sum(mask) # segment area computation

                # bbox computation for a segment
                hor = np.sum(mask, axis=0)
                hor_idx = np.nonzero(hor)[0]
                x = hor_idx[0]
                width = hor_idx[-1] - x + 1
                vert = np.sum(mask, axis=1)
                vert_idx = np.nonzero(vert)[0]
                y = vert_idx[0]
                height = vert_idx[-1] - y + 1
                bbox = [int(x), int(y), int(width), int(height)]

                segmInfo.append({"id": int(segmentId),
                                 "category_id": int(categoryId),
                                 "area": int(area),
                                 "bbox": bbox,
                                 "iscrowd": isCrowd})

            annotations.append({'image_id': imageId,
                                'file_name': outputFileName,
                                "segments_info": segmInfo})

            file_path = os.path.join(out_base_folder, channel.replace(kitti_sequence,""))
            os.makedirs(file_path,exist_ok=True)

            img = Image.open(os.path.join(path, imgName)).convert('RGB')

            sem = np.array(cat_pred[pan_pred])
            sem_tmp = sem.copy()
            sem_tmp[sem==255] = colors.shape[0] - 1
            sem_img = Image.fromarray(colors[sem_tmp])

            is_background = (sem < 11) | (sem == 255)
            #pan_pred = pan_pred.numpy()
            pan_pred = np.array(pan_pred)
            pan_pred[is_background] = 0

            sem_img = sem_img.resize((1226,370),Image.BICUBIC)

            out = Image.blend(img, sem_img, 1).convert(mode="RGBA")
            out.convert(mode="RGB").save(os.path.join(file_path, outputFileName))
            prog_bar.update()

    print("\nSaving the json file {}".format(out_base_json))
    d = {'images': images,
         'annotations': annotations,
         'categories': {}}
    with open(out_base_json, 'w') as f:
        json.dump(d, f, sort_keys=True, indent=4)


if __name__ == '__main__':
    main()

`

**Comment by mohan1914** [comment]
This happens when you feed the image with the original resolution (1226,370). We still have to figure out why the misalignment happens for non-training image resolution. We aim to update the repo in few days and would try to come up with a solution for this problem. However, meanwhile if you resize the image to the training image resolution (1280,384) before feeding it to the network the misalignment problem won't be there.

so in your code 
 img = cv2.resize(cv2.imread(os.path.join(path, imgName)),(1280,384))
 result = inference_detector(model, img, eval='panoptic')

if you make these changes. You will get the right predictions.

**Comment by GabrieleGalimberti-GaleSelector** [comment]
thank you very much!
I try it, then in a few hour I I let you know if I obtain the correct result.

**Comment by GabrieleGalimberti-GaleSelector** [comment]
> img = cv2.resize(cv2.imread(os.path.join(path, imgName)),(1280,384))

Is it necessary a resize with Bicubic interpolation and change color channels BGR2RGB ?

**Comment by mohan1914** [comment]
cv2 by default uses bilinear interpolation to resize and I think it should be enough. But you can always try bicubic as well.
No. The conversion from bgr2rgb happens during the normalization process with the given config files. So, you don't need to do it for the input image.

**Comment by GabrieleGalimberti-GaleSelector** [comment]
I see, thank you a lot! 
Now the result is ok for kitti configuration. 

![000346](https://user-images.githubusercontent.com/49124856/121268051-77402a00-c8bd-11eb-8bc5-bc575695bafb.png)

I'm trying Cityscapes configuration. Is It the output change exchangeable in the efficientPS_cityscapes.py file? for example, I need the same image size of output like kitti configuration.

**Comment by GabrieleGalimberti-GaleSelector** [comment]
for Cityscapes configuration is it need to resize image with size (1024,2048) ?

**Comment by mohan1914** [comment]
yes

**Comment by GabrieleGalimberti-GaleSelector** [comment]
Thank you very much!
